cmd/gpu_plugin/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,403 @@
+	rm := resourceManager{
+		nodeName:  os.Getenv("NODE_NAME"),
+		clientset: clientset,
+	}
+
+	klog.Info("GPU device plugin resource manager enabled")
+
+	return &rm
+}
+
+// reallocateWithFractionalResources runs the fractional resource logic.
+func (rm *resourceManager) reallocateWithFractionalResources(allocateResponse *pluginapi.AllocateResponse) error {
+	var err error
+	if err = reallocateInputCheck(allocateResponse); err != nil {
+		// it is better to give some gpu devices and return nil, than halt the whole cluster into unexpected admission errors
+		return nil
+	}
+
+	podCandidate, err := rm.findAllocationPodCandidate()
+	if err == errRetry {
+		klog.Warning("retrying POD resolving after sleeping")
+		time.Sleep(retryTimeout)
+		podCandidate, err = rm.findAllocationPodCandidate()
+	}
+
+	if err != nil {
+		if err != errZeroPending {
+			klog.Error("allocation candidate not found, perhaps the GPU scheduler extender is not called, err:" +
+				err.Error())
+		}
+		// it is better to give some gpu devices and return nil, than halt the whole cluster into unexpected admission errors
+		return nil
+	}
+
+	pod := podCandidate.pod
+	cards := containerCards(pod, podCandidate.allocatedContainers)
+	logPodInfo(pod, cards)
+
+	cresp := allocateResponse.ContainerResponses[0]
+	// select gpus for the allocation. the cards may differ from kubelet given deviceIDs!
+	for _, cardName := range cards {
+		rm.addDevicesAndMounts(cresp, cardName)
+	}
+	return nil
+}
+
+func reallocateInputCheck(allocateResponse *pluginapi.AllocateResponse) error {
+	// so far kubelet calls allocate for each container separately. If that changes, we need to refine our logic.
+	if len(allocateResponse.ContainerResponses) != 1 {
+		klog.Error(errMultiContainer.Error())
+		return errMultiContainer
+	}
+
+	for _, cresp := range allocateResponse.ContainerResponses {
+		if cresp.Annotations == nil {
+			klog.Error(errNoAnnotations.Error())
+			return errNoAnnotations
+		}
+
+		// if there is a device for the monitoring resource, skip the post allocation, the pod gets full access to all cards
+		devices := cresp.Annotations["INTEL_PLUGIN_DEVICES"]
+		if strings.Contains(devices, monitorID) {
+			return errMonitoringDevice // intentionally not printing anything
+		}
+	}
+
+	return nil
+}
+
+// findAllocationPodCandidate tries to find the best allocation candidate pod, which must be:
+//  -pending for this node
+//  -using GPU resources in its spec
+//  -is found via grpc service with unallocated GPU devices
+// returns:
+//  -the candidate pod struct pointer and no error, or
+//  -errRetry if unsuccessful, but there is perhaps hope of trying again with better luck
+//  -errZeroPending if no pending pods exist anymore (which is fine)
+//  -any grpc communication errors
+func (rm *resourceManager) findAllocationPodCandidate() (*podCandidate, error) {
+	// get map of pending pods for this node
+	pendingPods, err := rm.getNodePendingGPUPods()
+	if err != nil {
+		klog.Error(err.Error())
```

Comment:
Better log errors in the place where they are handled.

Reply:
done